From ba1384ee46845b992889d616896afb9b4c1d45d2 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Fri, 9 Apr 2021 23:19:41 -0500
Subject: [PATCH 070/196] Add SCRIPT_TAPROOT_KEY_SPEND_PATH script flag (#131)

Ensures that signatures are Schnorr signatures (64 byte long + sig hash flag) and use the BIP341 sighash.

This flag is required for verifying signatures for the Taproot key spend path (spending without revealing the script).

It is never set during script execution, only during signature verification.
---
 src/script/script_error.cpp          |   4 +
 src/script/script_error.h            |   5 ++
 src/script/script_flags.h            |   5 ++
 src/script/sigencoding.cpp           |  10 +++
 src/test/checkdatasig_tests.cpp      |   2 +
 src/test/schnorr_tests.cpp           |   2 +
 src/test/script_tests.cpp            |   4 +
 src/test/sigencoding_tests.cpp       | 130 ++++++++++++++++++---------
 src/test/txvalidationcache_tests.cpp |   9 +-
 9 files changed, 125 insertions(+), 46 deletions(-)

diff --git a/src/script/script_error.cpp b/src/script/script_error.cpp
index 6cfdb0a28..8e0ec4d45 100644
--- a/src/script/script_error.cpp
+++ b/src/script/script_error.cpp
@@ -110,6 +110,10 @@ std::string ScriptErrorString(const ScriptError serror) {
             return "Validation resources exceeded (SigChecks)";
         case ScriptError::INVALID_OP_SCRIPTTYPE:
             return "Marker opcode OP_SCRIPTTYPE cannot be executed";
+        case ScriptError::TAPROOT_KEY_SPEND_MUST_USE_BIP341_SIGHASH:
+            return "Taproot key spend signatures must use SIGHASH_BIP341";
+        case ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG:
+            return "Taproot key spend signature must be Schnorr";
         case ScriptError::UNKNOWN:
         case ScriptError::ERROR_COUNT:
         default:
diff --git a/src/script/script_error.h b/src/script/script_error.h
index 159682bf7..0a5bdc4d1 100644
--- a/src/script/script_error.h
+++ b/src/script/script_error.h
@@ -89,6 +89,11 @@ enum class ScriptError {
        marker */
     INVALID_OP_SCRIPTTYPE,
 
+    /* Key spend path for Taproot must use BIP341 sighash */
+    TAPROOT_KEY_SPEND_MUST_USE_BIP341_SIGHASH,
+    /* Key spend path for Taproot must Schnorr signatures */
+    TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG,
+
     ERROR_COUNT,
 };
 
diff --git a/src/script/script_flags.h b/src/script/script_flags.h
index faa2c8a5e..2a5ac96fb 100644
--- a/src/script/script_flags.h
+++ b/src/script/script_flags.h
@@ -11,6 +11,11 @@
 enum {
     SCRIPT_VERIFY_NONE = 0,
 
+    // Require BIP341 sighash and Schnorr signatures for key spend path
+    // (spending without revealing the script) in Taproot.
+    // Only used for checking signatures, not during script execution.
+    SCRIPT_TAPROOT_KEY_SPEND_PATH = (1U << 0),
+
     // Discourage use of NOPs reserved for upgrades (NOP1-10)
     //
     // Provided so that nodes can avoid accepting or mining transactions
diff --git a/src/script/sigencoding.cpp b/src/script/sigencoding.cpp
index 9d27d892c..33e298089 100644
--- a/src/script/sigencoding.cpp
+++ b/src/script/sigencoding.cpp
@@ -160,6 +160,11 @@ static bool IsSchnorrSig(const slicedvaltype &sig) {
 static bool CheckRawECDSASignatureEncoding(const slicedvaltype &sig,
                                            uint32_t flags,
                                            ScriptError *serror) {
+    if (flags & SCRIPT_TAPROOT_KEY_SPEND_PATH) {
+        // In Taproot key spend path, only Schnorr signatures are allowed.
+        return set_error(serror,
+                         ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG);
+    }
     if (IsSchnorrSig(sig)) {
         // In an ECDSA-only context, 64-byte signatures are forbidden.
         return set_error(serror, ScriptError::SIG_BADLENGTH);
@@ -216,6 +221,11 @@ static bool CheckSighashEncoding(const valtype &vchSig, uint32_t flags,
     }
 
     SigHashType sigHashType = GetHashType(vchSig);
+    // For Taproot key spend path, using BIP341 sighash is required
+    if ((flags & SCRIPT_TAPROOT_KEY_SPEND_PATH) && !sigHashType.hasBIP341()) {
+        return set_error(
+            serror, ScriptError::TAPROOT_KEY_SPEND_MUST_USE_BIP341_SIGHASH);
+    }
     bool usesForkId = sigHashType.hasForkId() || sigHashType.hasBIP341();
     bool forkIdEnabled = flags & SCRIPT_ENABLE_SIGHASH_FORKID;
     if (!forkIdEnabled && usesForkId) {
diff --git a/src/test/checkdatasig_tests.cpp b/src/test/checkdatasig_tests.cpp
index c476c332d..e0a71c7da 100644
--- a/src/test/checkdatasig_tests.cpp
+++ b/src/test/checkdatasig_tests.cpp
@@ -140,6 +140,8 @@ BOOST_AUTO_TEST_CASE(checkdatasig_test) {
     MMIXLinearCongruentialGenerator lcg;
     for (int i = 0; i < 4096; i++) {
         uint32_t flags = lcg.next();
+        // Flag cannot occur in script execution, only in Taproot key spend
+        flags &= ~SCRIPT_TAPROOT_KEY_SPEND_PATH;
 
         // Since strict encoding is enforced, hybrid keys are invalid.
         CheckError(flags, {{}, message, pubkeyH}, script,
diff --git a/src/test/schnorr_tests.cpp b/src/test/schnorr_tests.cpp
index 10b5e1e77..e929df1b0 100644
--- a/src/test/schnorr_tests.cpp
+++ b/src/test/schnorr_tests.cpp
@@ -92,6 +92,8 @@ BOOST_AUTO_TEST_CASE(opcodes_random_flags) {
     MMIXLinearCongruentialGenerator lcg(1234);
     for (int i = 0; i < 4096; i++) {
         uint32_t flags = lcg.next();
+        // Flag cannot occur in script execution, only in Taproot key spend
+        flags &= ~SCRIPT_TAPROOT_KEY_SPEND_PATH;
 
         const bool hasForkId = (flags & SCRIPT_ENABLE_SIGHASH_FORKID) != 0;
         const SigHashType sigHashType = SigHashType().withAlgorithm(
diff --git a/src/test/script_tests.cpp b/src/test/script_tests.cpp
index 6c379e7b5..14a50a7bd 100644
--- a/src/test/script_tests.cpp
+++ b/src/test/script_tests.cpp
@@ -95,6 +95,10 @@ static ScriptErrorDesc script_errors[] = {
     {ScriptError::INVALID_BIT_RANGE, "BIT_RANGE"},
     {ScriptError::INVALID_NUM2BIN_SIZE, "INVALID_NUM2BIN_SIZE"},
     {ScriptError::INVALID_OP_SCRIPTTYPE, "INVALID_OP_SCRIPTTYPE"},
+    {ScriptError::TAPROOT_KEY_SPEND_MUST_USE_BIP341_SIGHASH,
+     "TAPROOT_KEY_SPEND_MUST_USE_BIP341_SIGHASH"},
+    {ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG,
+     "TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG"},
 };
 
 static std::string FormatScriptError(ScriptError err) {
diff --git a/src/test/sigencoding_tests.cpp b/src/test/sigencoding_tests.cpp
index e40d7b64b..4033d5582 100644
--- a/src/test/sigencoding_tests.cpp
+++ b/src/test/sigencoding_tests.cpp
@@ -21,11 +21,17 @@ static valtype SignatureWithHashType(valtype vchSig, SigHashType sigHash) {
 static void CheckSignatureEncodingWithSigHashType(const valtype &vchSig,
                                                   uint32_t flags) {
     ScriptError err = ScriptError::OK;
-    BOOST_CHECK(CheckDataSignatureEncoding(vchSig, flags, &err));
 
     const bool hasForkId = (flags & SCRIPT_ENABLE_SIGHASH_FORKID) != 0;
+    // isTRKeySpend requires Schnorr sigs and BIP341 sighash
+    const bool isTRKeySpend = (flags & SCRIPT_TAPROOT_KEY_SPEND_PATH) != 0;
     const bool is64 = (vchSig.size() == 64);
 
+    // isTRKeySpend is never true for OP_CHECKDATASIG, but we test it anyway to
+    // make sure we don't introduce weird behavior
+    BOOST_CHECK_EQUAL(CheckDataSignatureEncoding(vchSig, flags, &err),
+                      !isTRKeySpend || is64);
+
     std::vector<BaseSigHashType> allBaseTypes{
         BaseSigHashType::ALL, BaseSigHashType::NONE, BaseSigHashType::SINGLE};
 
@@ -47,13 +53,18 @@ static void CheckSignatureEncodingWithSigHashType(const valtype &vchSig,
     }
 
     for (const SigHashType &sigHash : sigHashes) {
-        // Check the signature with the proper forkid flag.
+        // isTRKeySpend requires Schnorr sigs and BIP341 sighash
+        const bool expectedValid =
+            isTRKeySpend ? is64 && sigHash.hasBIP341() : true;
         valtype validSig = SignatureWithHashType(vchSig, sigHash);
-        BOOST_CHECK(CheckTransactionSignatureEncoding(validSig, flags, &err));
-        BOOST_CHECK_EQUAL(!is64, CheckTransactionECDSASignatureEncoding(
-                                     validSig, flags, &err));
-        BOOST_CHECK_EQUAL(is64, CheckTransactionSchnorrSignatureEncoding(
-                                    validSig, flags, &err));
+        BOOST_CHECK_EQUAL(expectedValid, CheckTransactionSignatureEncoding(
+                                             validSig, flags, &err));
+        BOOST_CHECK_EQUAL(
+            !is64 && expectedValid,
+            CheckTransactionECDSASignatureEncoding(validSig, flags, &err));
+        BOOST_CHECK_EQUAL(
+            is64 && expectedValid,
+            CheckTransactionSchnorrSignatureEncoding(validSig, flags, &err));
 
         // If we have strict encoding, we prevent the use of undefined flags.
         std::array<SigHashType, 5> undefSigHashes{
@@ -67,15 +78,21 @@ static void CheckSignatureEncodingWithSigHashType(const valtype &vchSig,
             valtype undefSighash = SignatureWithHashType(vchSig, undefSigHash);
             BOOST_CHECK(
                 !CheckTransactionSignatureEncoding(undefSighash, flags, &err));
-            BOOST_CHECK(err == ScriptError::SIG_HASHTYPE);
-            BOOST_CHECK(!CheckTransactionECDSASignatureEncoding(
-                                  undefSighash, flags, &err));
-            BOOST_CHECK(err == (is64 ? ScriptError::SIG_BADLENGTH
-                                        : ScriptError::SIG_HASHTYPE));
-            BOOST_CHECK(!CheckTransactionSchnorrSignatureEncoding(
-                                  undefSighash, flags, &err));
-            BOOST_CHECK(err == (!is64 ? ScriptError::SIG_NONSCHNORR
-                                        : ScriptError::SIG_HASHTYPE));
+            BOOST_CHECK_EQUAL(
+                err, isTRKeySpend && !is64
+                         ? ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG
+                         : ScriptError::SIG_HASHTYPE);
+            BOOST_CHECK(!CheckTransactionECDSASignatureEncoding(undefSighash,
+                                                                flags, &err));
+            BOOST_CHECK_EQUAL(
+                err, isTRKeySpend
+                         ? ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG
+                     : is64 ? ScriptError::SIG_BADLENGTH
+                            : ScriptError::SIG_HASHTYPE);
+            BOOST_CHECK(!CheckTransactionSchnorrSignatureEncoding(undefSighash,
+                                                                  flags, &err));
+            BOOST_CHECK_EQUAL(err, !is64 ? ScriptError::SIG_NONSCHNORR
+                                         : ScriptError::SIG_HASHTYPE);
         }
 
         // If we check strict encoding, then invalid forkid is an error.
@@ -85,20 +102,29 @@ static void CheckSignatureEncodingWithSigHashType(const valtype &vchSig,
 
         BOOST_CHECK(
             !CheckTransactionSignatureEncoding(invalidSig, flags, &err));
-        BOOST_CHECK(err == (hasForkId ? ScriptError::MUST_USE_FORKID
-                                        : ScriptError::ILLEGAL_FORKID));
+        BOOST_CHECK_EQUAL(
+            err,
+            isTRKeySpend
+                ? (is64 ? ScriptError::TAPROOT_KEY_SPEND_MUST_USE_BIP341_SIGHASH
+                        : ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG)
+                : (hasForkId ? ScriptError::MUST_USE_FORKID
+                             : ScriptError::ILLEGAL_FORKID));
         BOOST_CHECK(
             !CheckTransactionECDSASignatureEncoding(invalidSig, flags, &err));
-        BOOST_CHECK(err == (is64
-                                ? ScriptError::SIG_BADLENGTH
-                                : hasForkId ? ScriptError::MUST_USE_FORKID
-                                            : ScriptError::ILLEGAL_FORKID));
+        BOOST_CHECK_EQUAL(
+            err, isTRKeySpend
+                     ? ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG
+                 : is64      ? ScriptError::SIG_BADLENGTH
+                 : hasForkId ? ScriptError::MUST_USE_FORKID
+                             : ScriptError::ILLEGAL_FORKID);
         BOOST_CHECK(
             !CheckTransactionSchnorrSignatureEncoding(invalidSig, flags, &err));
-        BOOST_CHECK(err == (!is64
-                                ? ScriptError::SIG_NONSCHNORR
-                                : hasForkId ? ScriptError::MUST_USE_FORKID
-                                            : ScriptError::ILLEGAL_FORKID));
+        BOOST_CHECK_EQUAL(
+            err, !is64 ? ScriptError::SIG_NONSCHNORR
+                 : isTRKeySpend
+                     ? ScriptError::TAPROOT_KEY_SPEND_MUST_USE_BIP341_SIGHASH
+                 : hasForkId ? ScriptError::MUST_USE_FORKID
+                             : ScriptError::ILLEGAL_FORKID);
     }
 }
 
@@ -195,6 +221,7 @@ BOOST_AUTO_TEST_CASE(checksignatureencoding_test) {
         uint32_t flags = lcg.next();
 
         ScriptError err = ScriptError::OK;
+        const bool isTRKeySpend = (flags & SCRIPT_TAPROOT_KEY_SPEND_PATH) != 0;
 
         // Empty sig is always validly encoded.
         BOOST_CHECK(CheckDataSignatureEncoding({}, flags, &err));
@@ -210,22 +237,29 @@ BOOST_AUTO_TEST_CASE(checksignatureencoding_test) {
 
         // We enforce low S, therefore high S sigs are rejected.
         BOOST_CHECK(!CheckDataSignatureEncoding(highSSig, flags, &err));
-        BOOST_CHECK(err == ScriptError::SIG_HIGH_S);
+        BOOST_CHECK_EQUAL(
+            err, isTRKeySpend
+                     ? ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG
+                     : ScriptError::SIG_HIGH_S);
 
         for (const valtype &nonDERSig : nonDERSigs) {
             // If we get any of the dersig flags, the non canonical dersig
             // signature fails.
-            BOOST_CHECK(
-                !CheckDataSignatureEncoding(nonDERSig, flags, &err));
-            BOOST_CHECK(err == ScriptError::SIG_DER);
+            BOOST_CHECK(!CheckDataSignatureEncoding(nonDERSig, flags, &err));
+            BOOST_CHECK_EQUAL(
+                err, isTRKeySpend
+                         ? ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG
+                         : ScriptError::SIG_DER);
         }
 
         for (const valtype &nonDERSig : nonParsableSigs) {
             // If we get any of the dersig flags, the high S but non dersig
             // signature fails.
-            BOOST_CHECK(
-                !CheckDataSignatureEncoding(nonDERSig, flags, &err));
-            BOOST_CHECK(err == ScriptError::SIG_DER);
+            BOOST_CHECK(!CheckDataSignatureEncoding(nonDERSig, flags, &err));
+            BOOST_CHECK_EQUAL(
+                err, isTRKeySpend
+                         ? ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG
+                         : ScriptError::SIG_DER);
         }
     }
 }
@@ -381,22 +415,29 @@ BOOST_AUTO_TEST_CASE(checkschnorr_test) {
     MMIXLinearCongruentialGenerator lcg;
     for (int i = 0; i < 4096; i++) {
         uint32_t flags = lcg.next();
-
-        const bool hasForkId = (flags & SCRIPT_ENABLE_SIGHASH_FORKID) != 0;
+        const bool isTRKeySpend = (flags & SCRIPT_TAPROOT_KEY_SPEND_PATH) != 0;
+
+        SigHashType sigHashType = SigHashType();
+        if (isTRKeySpend) {
+            // Prevent invalid flag combinations
+            flags |= SCRIPT_ENABLE_SIGHASH_FORKID;
+            sigHashType = sigHashType.withBIP341();
+        } else if (flags & SCRIPT_ENABLE_SIGHASH_FORKID) {
+            sigHashType = sigHashType.withForkId();
+        }
 
         ScriptError err = ScriptError::OK;
-        valtype DER65_hb = SignatureWithHashType(
-            DER64, SigHashType().withAlgorithm(hasForkId ? SIGHASH_FORKID
-                                                         : SIGHASH_LEGACY));
-        valtype Zero65_hb = SignatureWithHashType(
-            Zero64, SigHashType().withAlgorithm(hasForkId ? SIGHASH_FORKID
-                                                          : SIGHASH_LEGACY));
+        valtype DER65_hb = SignatureWithHashType(DER64, sigHashType);
+        valtype Zero65_hb = SignatureWithHashType(Zero64, sigHashType);
 
         BOOST_CHECK(CheckDataSignatureEncoding(DER64, flags, &err));
         BOOST_CHECK(CheckTransactionSignatureEncoding(DER65_hb, flags, &err));
         BOOST_CHECK(
             !CheckTransactionECDSASignatureEncoding(DER65_hb, flags, &err));
-        BOOST_CHECK(err == ScriptError::SIG_BADLENGTH);
+        BOOST_CHECK_EQUAL(
+            err, isTRKeySpend
+                     ? ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG
+                     : ScriptError::SIG_BADLENGTH);
         BOOST_CHECK(
             CheckTransactionSchnorrSignatureEncoding(DER65_hb, flags, &err));
 
@@ -404,7 +445,10 @@ BOOST_AUTO_TEST_CASE(checkschnorr_test) {
         BOOST_CHECK(CheckTransactionSignatureEncoding(Zero65_hb, flags, &err));
         BOOST_CHECK(
             !CheckTransactionECDSASignatureEncoding(Zero65_hb, flags, &err));
-        BOOST_CHECK(err == ScriptError::SIG_BADLENGTH);
+        BOOST_CHECK_EQUAL(
+            err, isTRKeySpend
+                     ? ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG
+                     : ScriptError::SIG_BADLENGTH);
         BOOST_CHECK(
             CheckTransactionSchnorrSignatureEncoding(Zero65_hb, flags, &err));
     }
diff --git a/src/test/txvalidationcache_tests.cpp b/src/test/txvalidationcache_tests.cpp
index 4819f3f4e..d3ea8bfc6 100644
--- a/src/test/txvalidationcache_tests.cpp
+++ b/src/test/txvalidationcache_tests.cpp
@@ -466,9 +466,12 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, TestChain100Setup) {
 
         // This should be valid under all script flags that support our sighash
         // convention.
-        ValidateCheckInputsForAllFlags(
-            CTransaction(tx), SCRIPT_ENABLE_REPLAY_PROTECTION,
-            SCRIPT_ENABLE_SIGHASH_FORKID, true, 2);
+        // If the SCRIPT_TAPROOT_KEY_SPEND_PATH is set (which never occurs
+        // during script execution), the tx fails.
+        ValidateCheckInputsForAllFlags(CTransaction(tx),
+                                       SCRIPT_ENABLE_REPLAY_PROTECTION |
+                                           SCRIPT_TAPROOT_KEY_SPEND_PATH,
+                                       SCRIPT_ENABLE_SIGHASH_FORKID, true, 2);
 
         {
             // Try checking this valid transaction with sigchecks limiter
-- 
2.39.5 (Apple Git-154)

